common/client-libs/gateway-client/src/bandwidth.rs

```diff
@@ -0,0 +1,190 @@
+            .map_err(|_| GatewayClientError::InvalidEthereumPrivateKey)?;
+
+        Ok(BandwidthController {
+            contract,
+            eth_private_key,
+        })
+    }
+
+    #[cfg(feature = "coconut")]
+    pub async fn prepare_coconut_credential(
+        &self,
+    ) -> Result<coconut_interface::Credential, GatewayClientError> {
+        let verification_key = obtain_aggregate_verification_key(&self.validator_endpoints).await?;
+
+        let bandwidth_credential =
+            obtain_signature(&self.identity.to_bytes(), &self.validator_endpoints).await?;
+        // the above would presumably be loaded from a file
+
+        // the below would only be executed once we know where we want to spend it (i.e. which gateway and stuff)
+        Ok(prepare_for_spending(
+            &self.identity.to_bytes(),
+            &bandwidth_credential,
+            &verification_key,
+        )?)
+    }
+
+    #[cfg(not(feature = "coconut"))]
+    pub async fn prepare_token_credential(
+        &self,
+        gateway_identity: PublicKey,
+    ) -> Result<TokenCredential, GatewayClientError> {
+        let mut rng = OsRng;
+
+        let kp = identity::KeyPair::new(&mut rng);
+
+        let verification_key = *kp.public_key();
+        let signed_verification_key = kp.private_key().sign(&verification_key.to_bytes());
+        self.buy_token_credential(verification_key, signed_verification_key)
+            .await?;
+
+        let message: Vec<u8> = verification_key
+            .to_bytes()
+            .iter()
+            .chain(gateway_identity.to_bytes().iter())
+            .copied()
+            .collect();
+        let signature = kp.private_key().sign(&message);
+
+        Ok(TokenCredential::new(
+            verification_key,
+            gateway_identity,
+            BANDWIDTH_VALUE,
+            signature,
+        ))
+    }
+
+    #[cfg(not(feature = "coconut"))]
+    pub async fn buy_token_credential(
+        &self,
+        verification_key: PublicKey,
+        signed_verification_key: identity::Signature,
+    ) -> Result<(), GatewayClientError> {
+        // 0 means a transaction failure, 1 means success
+        if Some(U64::from(0))
+            == self
+                .contract
+                .signed_call_with_confirmations(
+                    ETH_BURN_FUNCTION_NAME,
+                    (
+                        U256::from(TOKENS_TO_BURN),
+                        U256::from(&verification_key.to_bytes()),
+                        Bytes(signed_verification_key.to_bytes().to_vec()),
+                    ),
+                    Options::default(),
+                    1,
+                    &self.eth_private_key,
+                )
+                .await?
+                .status
+        {
+            Err(GatewayClientError::BurnTokenError(
+                web3::Error::InvalidResponse(String::from("Transaction status is 0 (failure)")),
```

Comment:
This would be a good thing, in theory, but from what I've observed, the it looks like the logs are populated when the transaction is a success. On error, it's actually just an empty vector of `Log`... Maybe this is just for the error I've encountered (when my eth account didn't have any ERC20 tokens) so I'll include the logs in the error message, but I wouldn't get my hopes up.